Add achievements and leaderboard cookbook (#9678)
flutter/website: src/cookbook/games/achievements-leaderboard.md

---
title: Add achievements and leaderboards to your mobile game
description: >
  How to use the games_services plugin to add functionality to your game.
---

Gamers have various motivations for playing games. In broad strokes,
there are four major motivations: 
[immersion, achievement, cooperation, and competition](https://meditations.metavert.io/p/game-player-motivations).

> No matter the game you build, some players want to *achieve* in it. This
> could be trophies won or secrets unlocked. Some players want to
> *compete* in it. This could be hitting high scores or accomplishing
> speedruns. These two ideas map to the concepts of *achievements* and
> *leaderboards*.

Reviewer: No textual changes here, just breaking up the sentences so any future edits/diffs are easier to review. 

```suggestion
No matter the game you build, some players want to *achieve* in it.
This could be trophies won or secrets unlocked.
Some players want to *compete* in it.
This could be hitting high scores or accomplishing speedruns.
These two ideas map to the concepts of *achievements* and *leaderboards*.
```

![A simple graphic representing the four types of motivation explained above](/assets/images/docs/cookbook/types-of-gamer-motivations.png)

> Ecosystems such as the App Store and Google Play provide centralized
> services for achievements and leaderboards. Players have achievements
> from all the games in one place and developers don't need to
> re-implement them for every game.

Reviewer: When reading, I wasn't sure what "all the games" meant. Would something like "players **can view** achievements from **all their** games..." make sense?

```suggestion
Ecosystems such as the App Store and Google Play provide
centralized services for achievements and leaderboards.
Players can view achievements from all their games in one place and
developers don't need to re-implement them for every game.
```

> This recipe demonstrates how to use the `games_services` package to add

Author: Fixed

achievements and leaderboard functionality to your mobile game.

## 1. Enable platform services

> To enable games services, set up *Game Center* on iOS and *Google Play
> Games Services* on Android.

Reviewer: Semantic breaks to make future diffs/changes easier to review.

```suggestion
To enable games services, set up *Game Center* on iOS and
*Google Play Games Services* on Android.
```

> ### iOS

Reviewer: From a brief glance at the [GameKit docs](https://developer.apple.com/documentation/gamekit/), it seems it supports macOS. Would these instructions work for macOS and the Mac App Store too?

If so, might be worth calling this section "iOS and macOS" and updating the various mentions, or at least calling out somewhere in the section that these instruction can be applied to macOS too.

Author: I haven't tested this, so I'd rather underpromise than overpromise. I _think_ GameKit behaves a bit differently on macOS than on iOS, so it needs further investigation. I'm making a note to self.

Reviewer: Sounds good, expanding things later is great too :)

To enable Game Center (GameKit) on iOS:

1.  Open your Flutter project in Xcode.
    Open `ios/Runner.xcworkspace`

2.  Select the root **Runner** project.

3.  Go to the **Signing & Capabilities** tab.

4.  Click the `+` button to add **Game Center** as a capability.

5.  Close Xcode.

> 6.  If you haven't already, register your game 

Reviewer: I'm not super familiar with iOS development or deployment, but does this require being part of the Apple developer program?

If so, can we add a mention here or the the beginning of the iOS section that this a prerequisite and link to the relevant page. I believe we have some examples of this on the site already.

Author: I feel that might be going too far? Not even the page you're mentioning (https://docs.flutter.dev/deployment/ios) talks about the Apple developer program sign up process. As much as I like to inline knowledge so people don't need to go hunting for it elsewhere, I think that teaching people here that, for an iOS game, they need an Apple Developer account, is excessive.

That said, I don't _mind_ putting it here. In the end, it's a judgement call. I'd lean towards not mentioning such things, that's all.

Reviewer: That makes sense. I'm fine leaving the note out. Thanks!

    game. It only publishes the achievements and leaderboard. Once a
    leaderboard, for example, is published this way, it cannot be
    unpublished.

6.  Go to **Play Games Services** **→ Setup and management →
    Configuration → Credentials**.

7.  Find the **Get resources** button. It returns an XML file with the
    Play Games Services IDs.

    ```xml
    <!-- THIS IS JUST AN EXAMPLE -->
    <?xml version="1.0" encoding="utf-8"?>
    <resources>
        <!--app_id-->
        <string name="app_id" translatable="false">424242424242</string>
        <!--package_name-->
        <string name="package_name" translatable="false">dev.flutter.tictactoe</string>
        <!--achievement First win-->
        <string name="achievement_first_win" translatable="false">sOmEiDsTrInG</string>
        <!--leaderboard Highest Score-->
        <string name="leaderboard_highest_score" translatable="false">sOmEiDsTrInG</string>
    </resources>
    ```

8.  Add a file at `android/app/src/main/res/values/games-ids.xml`
    containing the XML you received in the previous step.

## 2. Sign in to the game service

Now that you have set up *Game Center* and *Play Games Services*, and

> have your achievement & leaderboard IDs ready, it's finally Dart time.

Reviewer: I love Dart time!

> 1.  Add a dependency on the [`games_services` package](https://pub.dev/packages/games_services).

Reviewer: ```suggestion
1.  Add a dependency on the [`games_services` package]({{site.pub-pkg}}/games_services).
```

>     ```shell
>     flutter pub add games_services
>     ```

Reviewer: To be consistent with elsewhere on the site:

````suggestion
    ```terminal
    $ flutter pub add games_services
    ```
````

Author: Fixed

2.  Before you can do anything else, you have to sign the player into
    the game service.

    ```dart
    try {
      await GamesServices.signIn();
    } on PlatformException catch (e) {
      // ... deal with failures ...
    }
    ```

The sign in happens in the background. It takes several seconds, so
don't call `signIn()` before `runApp()` or the players will be forced to
stare at a blank screen every time they start your game.

The API calls to the `games_services` API can fail for a multitude of
reasons. Therefore, every call should be wrapped in a try-catch block as
in the previous example. The rest of this recipe omits exception
handling for clarity.

> {{site.alert.note}}

Reviewer: The tip aside might draw more attention to this :) 

```suggestion
{{site.alert.tip}}
```

  It's a good practice to create a controller. This would be a
  `ChangeNotifier`, a bloc, or some other piece of logic that wraps around
  the raw functionality of the `games_services` plugin.
{{site.alert.end}}

## 3. Unlock achievements

1.  Register achievements in Google Play Console and App Store Connect,
    and take note of their IDs. Now you can award any of those
    achievements from your Dart code:

    ```dart
    await GamesServices.unlock(
      achievement: Achievement(
        androidID: 'your android id',
        iOSID: 'your ios id',
      ),
    );
    ```

    The player's account on Google Play Games or Apple Game Center now
    lists the achievement.

2.  To display the achievements UI from your game, call the
    `games_services` API:

    ```dart
    await GamesServices.showAchievements();
    ```

    This displays the platform achievements UI as an overlay on your game.

3.  To display the achievements in your own UI, use

>     `GamesServices.loadAchievements()`.

Reviewer: Consider linking this to the package's API docs:

```suggestion
    [`GamesServices.loadAchievements()`][].
    
[`GamesServices.loadAchievements()`]: {{site.pub-api}}/games_services/latest/games_services/GamesServices/loadAchievements.html
```

## 4. Submit scores

When the player finishes a play-through, your game can submit the result
of that play session into one or more leaderboards.

For example, a platformer game like Super Mario can submit both the
final score and the time taken to complete the level, to two separate
leaderboards.

1.  In the first step, you registered a leaderboard in Google Play
    Console and App Store Connect, and took note of its ID. Using this
    ID, you can submit new scores for the player:

    ```dart
    await GamesServices.submitScore(
      score: Score(
        iOSLeaderboardID: 'some_id_from_app_store',
        androidLeaderboardID: 'sOmE_iD_fRoM_gPlAy',
        value: 100,
      ),
    );
    ```

    You don't need to check whether the new score is the player's
    highest. The platform game services handle that for you.

2.  To display the leaderboard as an overlay over your game, make the
    following call:

    ```dart
    await GamesServices.showLeaderboards(
      iOSLeaderboardID: "some_id_from_app_store",
      androidLeaderboardID: "sOmE_iD_fRoM_gPlAy",
    );
    ```

3.  If you want to display the leaderboard scores in your own UI, you

>     can fetch them with `GamesServices.loadLeaderboardScores()`.

Reviewer: Consider linking this to the package's API docs:

```suggestion
    can fetch them with [`GamesServices.loadLeaderboardScores()`][].
    
[`GamesServices.loadLeaderboardScores()`]: {{site.pub-api}}/games_services/latest/games_services/GamesServices/loadLeaderboardScores.html
```

> ## 5. Complete example

Reviewer: I'm not sure "Complete example" makes sense as the title of this step, perhaps something such as `5. Next steps` or similar?

Author: Good point, fixed

There's more to the `games_services` plugin. With this plugin, you can:

> -   get the player's icon, name or unique ID
> 
> -   save and load game states
> 
> -   sign out of the game service

Reviewer: ```suggestion
- Get the player's icon, name or unique ID
- Save and load game states
- Sign out of the game service
```

Some achievements can be incremental. For example: "You have collected

> all 10 pieces of the McGuffin.".

Reviewer: ```suggestion
all 10 pieces of the McGuffin."
```

Each game has different needs from game services.

> To start, you might want to create this controller:

Reviewer: Can you expand a bit more on "why" you might want to create this controller? Is it to encapsulate the previous steps?

Author: Ack

> ```dart

Reviewer: I love this fuller example!

Could you update this to use code excerpts and place the source in [`/examples/cookbook/games/achievements-leaderboard`](https://github.com/flutter/website/tree/main/examples/cookbook), or create an issue so I or someone else can follow up to do so? That way we can keep up to date with changes in the packages and make sure analysis passes with Dart changes.

Author: Yes, I'll create an issue. I'll probably pick up the issue myself, in a week or so. I just want to make this ready as soon as possible.
